---
title: "Annotate Cell Types and Save Individual Cell-Type ACTIONet Objects"
output:
  html_notebook: default
  pdf_document: default
---
```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = "/home/samsl/Pseudotemporal-GRNs")
```

```{r "Load Libraries", warning=FALSE, include=TRUE, results='hide'}
require(ACTIONet)
require(SingleCellExperiment)
require(ComplexHeatmap)
require(Matrix)
require(scater)
require(scran)
```

# Create SingleCellExperiment Object
If using an mtx file and metadata, create an sce object. If you already have an SCE, Seurat, or AnnData object, skip to the relevant section.
```{r "Create Single Cell Experiment"}
# Matrix and Metadata files
matrix.file.name = "data/Processed_10x_AD_data/filtered_count_matrix.mtx"
colMetadata.file.name = "data/Processed_10x_AD_data/filtered_column_metadata.txt"
idMapping.file.name = "data/Processed_10x_AD_data/id_mapping.csv"
rowNames.file.name = "data/Processed_10x_AD_data/filtered_gene_row_names.txt"

# Read in counts and metadata
filtered.counts <- readMM(matrix.file.name)
filtered.colMetadata <- read.delim(colMetadata.file.name)
filtered.colMetadata$pre.cluster <- factor(filtered.colMetadata$pre.cluster)

# Map cells to their samples
idMapping <- read.table(idMapping.file.name, header = TRUE, sep = ",", stringsAsFactors = FALSE)
filtered.colMetadata$Subject <- as.factor(unlist(lapply(filtered.colMetadata$projid, function(x) { idMapping$Subject[[match(x,idMapping$projid)]] })))

# Assign row names (genes) and column names (cell tags)
rownames(filtered.counts) <- readLines(rowNames.file.name)
colnames(filtered.counts) <- filtered.colMetadata$TAG

# Create tsne array
tsneDims <- filtered.colMetadata[c("tsne1", "tsne2")]

# Create SCE
colData <- filtered.colMetadata[c("projid","Subject","pre.cluster","broad.cell.type","Subcluster")]
sce <- SingleCellExperiment(assays = list(counts =filtered.counts), colData = colData)
reducedDim(sce) <- tsneDims
reducedDimNames(sce) <- "tsne"
rowData(sce)$hgnc_symbol <- rownames(sce)

```

## Processing
Normalize to log counts and perform batch correction with Harmony.
```{r "Reduce and Batch Correct"}
# Reduce and Batch Correct
sce <- reduce.and.batch.correct.sce.Harmony(sce, batch.vec = colData(sce)$Subject, reduced_dim = 35)
```

Add gene metadata from biomart, including  ensembl symbol and position on the chromosome.
```{r "Biomart Annotation"}
# Add Biomart Information to Genes
require(biomaRt)
mart=useMart(biomart="ensembl", dataset="hsapiens_gene_ensembl")

bm.attributes <- c("ensembl_gene_id", "start_position", "end_position", "chromosome_name")
sce <- getBMFeatureAnnos(sce, filters="hgnc_symbol",
                            attributes = bm.attributes,
                            dataset = "hsapiens_gene_ensembl")
```

### Load from processed SCE Object
```{r "Load from SCE"}
# SCE file name
sce.file.name <- "~/Pseudotemporal-GRNs/data/sce.rds"

sce <- readRDS(sce.file.name)
```

### Load from Seurat Object
```{r "Load from Seurat"}
# Seurat File Name
seurat.file.name <- ""

sce <- import.sce.from.Seurat(readRDS(file = seurat.file.name))
```

### Load from AnnData File
```{r "Load from AnnData"}
# AnnData File Name
anndata.file.name <- ""

sce <- import.sce.from.AnnData(file = anndata.file.name)
```

## Save
Save the SCE for future analyses.
```{r "Save SCE", result='hide'}
sce.file.name <- "data/sce.rds"
saveRDS(sce, file = sce.file.name)
```

# Annotate Cell Types
Run ACTIONet to create the ACTIONet object and allow for cell state discovery and cell type annotation.
```{r "Compute ACTIONet"}
ACTIONet.out <- run.ACTIONet(sce, k_max = 20, thread_no = 12)
```
Once the ACTIONet has been constructed, choose the set of markers for the cell types in your sample and run annotation using those markers.
```{r "Cell Type Annotation"}
# Select Marker Set
data("curatedMarkers_human")
markers <- curatedMarkers_human$Brain$PFC$major.celltypes$marker.genes

# Annotate cell types
ACTIONet.out <- annotate.cells.from.archetypes.using.markers(ACTIONet.out,
                                                             markers,
                                                             annotation.name = "Cell.Types")

# Plot to show cell types
png('img/ACTIONet_cell_types.png')
plot.ACTIONet(ACTIONet.out, labels = ACTIONet.out$annotations$Cell.Types$Labels, transparency.attr = ACTIONet.out$annotations$Cell.Types$Labels.confidence)
dev.off()
```
Save the full annotated ACTIONet object for later analyses.
```{r "Save annotated"}
# Full ACTIONet File
full.ACTIONet.file <- "data/allCellsACTIONet.rds"

# Save annotated full ACTIONet
saveRDS(ACTIONet.out, file = full.ACTIONet.file)
```

### Load Annotated ACTIONet
```{r "Load Annotated ACTIONet"}
# Full ACTIONet File
full.ACTIONet.file <- "~/Pseudotemporal-GRNs/data/allCellsACTIONet.rds"

ACTIONet.out <- readRDS(file = full.ACTIONet.file)
```

```{r "Identify disease enriched archetype"}
# Identify disease enriched archetype
data("nanoStringDB_human")
ADmarkers <- nanoStringDB_human$AD
enrichment.out <- t(geneset.enrichment.archetype(ACTIONet.out, ADmarkers))
png("All_cell_state_disease_association.png")
Heatmap(t(enrichment.out), cluster_columns=FALSE, cluster_rows=FALSE, name='Enrichment', column_title='Cell State', column_title_side='bottom')
```

```{r "Calculate disease scores"}
correlated.state <- which.max(enrichment.out[,2])

disease.score <- ACTIONet.out$unification.out$H.core[correlated.state,]
core.significance <- ACTIONet.out$unification.out$DE.core@assays[["significance"]]
associated.genes <- rownames(core.significance)[order(core.significance[,correlated.state], decreasing=TRUE)]
ACTIONet.out <- add.cell.annotations(ACTIONet.out, disease.score, annotation.name = "disease.score")
plot.ACTIONet.gradient(ACTIONet.out,disease.score)
```

```{r}
plot.ACTIONet <- function(ACTIONet.out, labels = NULL, transparency.attr = NULL, trans.z.threshold = -0.5, trans.fact = 3, 
	node.size = 1, CPal = ACTIONet.color.bank, add.text = FALSE, text.halo.width = 0.1, label.text.size = 0.8, 
    suppress.legend = FALSE, legend.pos = "bottomright", add.states = F, title = "", highlight = F) {
    
    node.size = node.size * 0.5
    
    if (is.igraph(ACTIONet.out)) 
        ACTIONet = ACTIONet.out else ACTIONet = ACTIONet.out$ACTIONet
    
    coors = cbind(V(ACTIONet)$x, V(ACTIONet)$y)

	if( (length(labels) == 1) & (is.character(labels)) ) {
		annotation_name = labels
	} else {
		annotation_name = "plot.ACTIONet-tmp"
	}	
    
	labels = preprocess.labels(ACTIONet.out, labels)

	if( !is.null(labels) & (highlight == T) ) {
		if( annotation_name == "plot.ACTIONet-tmp") {
			ACTIONet.out = add.cell.annotations(ACTIONet.out, cell.annotations = labels, annotation.name = annotation_name)
		}
		
		if( is.null(ACTIONet.out$annotations[[annotation_name]]$highlight) ) {
			print("Labels are not highlighted ... generating highlight on the fly")
			ACTIONet.out = highlight.annotations(ACTIONet.out, annotation.name = annotation_name)			
		}
		label.hightlight = ACTIONet.out$annotations[[annotation_name]]$highlight
		
		if( !is.null(transparency.attr) ) {
			print(class(transparency.attr))
			print("highlight=T while transparency.attr is not NULL. Overwriting transparency.attr with highlight values")
		}
		transparency.attr = label.hightlight$connectivity.scores		
	}
	
	if(is.null(labels)) {
        vCol = V(ACTIONet)$color
        Annot = NULL
	} else {
		Annot = names(labels)[match(sort(unique(labels)), labels)]		
		if(length(CPal) > 1) {
            Pal = CPal[1:length(Annot)]			
		} else {
            Pal = ggpubr::get_palette(CPal, length(Annot))
		}
        names(Pal) = Annot
        vCol = Pal[names(labels)]
	}



    if (!is.null(transparency.attr)) {
        #z = (transparency.attr - median(transparency.attr))/mad(transparency.attr)
        z = scale(transparency.attr)  
      beta = 1/(1 + exp(-trans.fact * (z - trans.z.threshold)))
        beta[z > trans.z.threshold] = 1
        beta = beta^trans.fact
        
        vCol.border = scales::alpha(colorspace::darken(vCol, 0.5), beta)
        vCol = scales::alpha(vCol, beta)
    } else {
        vCol.border = colorspace::darken(vCol, 0.5)
    }

    graphics::plot(coors[, c(1, 2)], pch = 21, cex = node.size, bg = vCol, col = vCol.border, axes = FALSE, xlab = "", ylab = "", main = title)

	if(add.states == T) {
		par(new=TRUE)
		
		M = as(ACTIONet.out$unification.out$C.core, 'sparseMatrix')
		cs = Matrix::colSums(M)
		M = scale(M, center = FALSE, scale = cs)
		
	    cell.Lab = grDevices::convertColor(color = t(col2rgb(vCol)/256), from = "sRGB", to = "Lab")	    
	    core.Lab = t(t(cell.Lab) %*% M)
	    core.colors = rgb(grDevices::convertColor(color = core.Lab, from = "Lab", to = "sRGB"))
		core.colors = colorspace::lighten(core.colors, 0.1)
		
		core.coors = t(t(ACTIONet.out$vis.out$coordinates) %*% M)		
	    
	    graphics::plot(core.coors, pch = 25, cex = 4*node.size, bg = core.colors, col = "#eeeeee", axes = FALSE, xlab = "", ylab = "")
	}    
    
    
    if ( add.text == T & (!is.null(Annot)) ) {
    	require(wordcloud)
        centroids = t(sapply(Annot, function(l) {
            idx = which(names(labels) == l)
            if(length(idx) == 1) {
				return(as.numeric(coors[idx, ]))
			} 

            sub.coors = coors[idx, ]
            anchor.coor = as.numeric(apply(sub.coors, 2, function(x) mean(x, trim = 0.5)))
#             sub.coors.sq = sub.coors^2
# 			norm.sq = Matrix::rowSums(sub.coors.sq)
# 			anchor.idx = which.min(sapply(1:nrow(sub.coors.sq), function(i) { 
# 				dd = norm.sq[i] + norm.sq - 2* sub.coors %*% sub.coors[i, ]
# 				mean.dist.sq = median(dd)
# 				return(mean.dist.sq)
# 			}))
            
                        
            # D = as.matrix(dist(sub.coors))
            # stats = Matrix::rowMeans(D)
            # anchor.idx = which.min(stats)
            
			# anchor.coor = as.numeric(sub.coors[anchor.idx, ])            
            
			return(anchor.coor)
        }))
        layout.labels(x = centroids[, 1], y = centroids[, 2], labels = Annot, col = colorspace::darken(Pal, 0.5), bg = "#eeeeee", r = text.halo.width, cex = label.text.size)
        #wordcloud::textplot(x = centroids[, 1], y = centroids[, 2], new = F, words = Annot, col = colorspace::darken(Pal, 0.5), bg = "#eeeeee") 
    }
    
    if ( (suppress.legend == FALSE) & !is.null(Annot) ) {
		#par(xpd = T, mar = par()$mar * c(1,1,1,1.1))
        #legend(legend.pos, legend = Annot, fill = Pal, cex = 0.5, bty = "n", inset=c(-0.1,0))
        legend(legend.pos, legend = Annot, fill = Pal, cex = 0.5, bty = "n")
    }    
}

```

```{r Show cell types associated with disease}
transparency.attr = log(disease.score + 0.0001*rnorm(length(disease.score)))
# trans.fact = 2
# trans.z.threshold = -1

        z = (transparency.attr - median(transparency.attr))/mad(transparency.attr)
        beta = 1/(1 + exp(-trans.fact * (z - trans.z.threshold)))
        beta[z > trans.z.threshold] = 1
        beta = beta^trans.fact

  plot(density(z))
  
plot.ACTIONet(ACTIONet.out, labels = ACTIONet.out$annotations$Cell.Types$Labels, transparency.attr = disease.score, trans.z.threshold = 0.5, trans.fact = 3)
```

## Save SCE and ACTIONet For Each Cell Type
We want to re-run ACTIONet for each subset of cells to learn archetypes within a cell type
```{r "Create ACTIONet for each cell type", result='hide'}
cell.type.file.dir = "data"

for (cell.type in names(markers)) {
  print(paste("Running ACTIONet for",cell.type,"cells..."))
  keep.cells <- colnames(sce)[which(ACTIONet.out$annotations$Cell.Types$Labels == cell.type)]
  if (length(keep.cells) == 0) {
    print(paste("No cells labeled as ",cell.type))
    next 
  }
  cell.type.sce = sce[,keep.cells]
  cell.type.ACTIONet = run.ACTIONet(cell.type.sce, k_max = 20, thread_no = 12)
  saveRDS(cell.type.sce, file = paste(cell.type.file.dir,"/",cell.type,"_sce.rds",sep=""))
  saveRDS(cell.type.ACTIONet, file = paste(cell.type.file.dir,"/",cell.type,"_ACTIONet.rds",sep=""))
}

```
Now we can perform the analyses in constructPseudotemporalOrdering.Rmd using each `[cell.type]_sce.rds` and `[cell.type]_ACTIONet.rds`.